### PR TITLE
Add clippy at workspace level

### PR DESCRIPTION
**Pull request by tcoratger**
## Description

This pull request adds the Clippy tool at the workspace level. By enabling Clippy at the workspace level, it will run on all projects within the workspace, providing consistent linting across the entire codebase.

## Changes Made

- Added Clippy as a development dependency in the workspace's `Cargo.toml` file.
- Configured Clippy through the `clippy.toml` file to customize linting rules and configurations.

### CODE REVIEW

File: Cargo.toml
Comment (onbjerg) [inline Cargo.toml:23]: unsure about these, defer to @prestwich and @mattsse if we should just keep this default or not

obviously some of them should be kept to allow us to remove some of the `#![warn]` and `#![deny]`s

**Review by onbjerg — APPROVED** [review]
sensible idea to me

File: Cargo.toml
Comment (prestwich) [reply Cargo.toml:23]: this is a big list to review. 

some of them, like `missing_const_for_fn` are definitely not set correctly.  we _do_ want to enforce `missing_const_for_fn`, in spite of the bug wrt `const fn` destructors

For nursery lints, the comment needs to be dated with when the list was reproduced here, and link to documentation of the full list of nursery lints. It needs to be maintained regularly or will fall out of date with the nursery.  This is the sort of annoying churn and manual overhead that I try to avoid and is why we stopped using nightly clippy in the first place

Overall, I am in favor of dropping the whole list and keeping defaults

File: crates/alloy/src/lib.rs
Comment (prestwich) [inline crates/alloy/src/lib.rs:6]: good. appreciate deleting these.

File: crates/json-rpc/src/response/mod.rs
Comment (prestwich) [inline crates/json-rpc/src/response/mod.rs:80]: this line too?

File: crates/provider/src/provider/with_block.rs
Comment (prestwich) [inline crates/provider/src/provider/with_block.rs:112]: can condense to `is_null()` while we're cleaning :)

File: Cargo.toml
Comment (tcoratger) [reply Cargo.toml:23]: @prestwich But if we keep only the default, this means the following list

```
rust.missing_debug_implementations = "warn"
rust.missing_docs = "warn"
rust.unreachable_pub = "warn"
rust.unused_must_use = "deny"
rust.rust_2018_idioms = "deny"
rustdoc.all = "warn"
```

and for example we omit `use_self` which was quite useful here or `missing_const_for_fn` that I can add to the warn list (my bad). As a ref, I used the same list as in reth.

File: crates/provider/src/provider/with_block.rs
Comment (tcoratger) [reply crates/provider/src/provider/with_block.rs]: fixed

File: Cargo.toml
Comment (tcoratger) [reply Cargo.toml:23]: @prestwich I've just updated by pushing `all` (all lints that are on by default (correctness, suspicious, style, complexity, perf))

I added 3 others as supplements which seem stable and quite useful compared to what we have in the codebase: 
```
missing_const_for_fn = "warn"
use_self = "warn"
option_if_let_else = "warn"
```

tell me what you think?

**Comment by prestwich** [comment]
will give this another pass in the morning

File: Cargo.toml
Comment (prestwich) [reply Cargo.toml:23]: Cool, this looks fine for me. Can merge once we re-apply the lints on latest main

**Comment by prestwich** [comment]
sorry, i thought i left a comment days ago. i'm good with the approach and will approve and merge once re-applied to main

**Comment by tcoratger** [comment]
> sorry, i thought i left a comment days ago. i'm good with the approach and will approve and merge once re-applied to main

Just did a rebase :)

**Pull request merged**
